### ISSUE TITLE
Tree Table implementation

### ISSUE THREAD

**Issue by mcarrano** [patternfly/patternfly-react#5016]
We would like to introduce tree table support based on the Cloud Platform implementation here: https://github.com/RedHatInsights/frontend-components/tree/master/packages/components/src/Components/TreeTable. You can see a live demo here: https://cloud.redhat.com/docs/api

Some questions to investigate include:

* Whether this is a new component or can be introduced as a new variant of the existing React table. Would like to inherit styling options from the existing table, but concerned about continued bloat of the Table component.
* Can we build using lightweight table approach that is currently being exposed (composable table) to alleviate size/complexity concerns?

There are currently known issues with responsive behavior. I have opened a corresponding design issue here to recommend an approach: https://github.com/patternfly/patternfly-design/issues/931

cc @evwilkin @mattnolting @karelhala

**Comment by karelhala** [comment]
The way how I'd like to see the table to be reimagined is that we could export original Table component with data driven approach from current `@patternfly/react-table` package, this package could use 2 new packages:
 * `@patternfly/react-table-lightweight` - basic HTML components, `Table`, `TableHead`, `TableBody`, `TableFooter`, `TableRow` and `TableCell`. These would be just dumb components without any real added behavior, no option to select rows, no sort capabilities, no expandable row and other more complex table behaviors. It would essentially be just table components with some styling.
 * `@patternfly/react-table-extensions` - all decorators and extensions to table including new `tableTreeView`. Importing for instance extension for collapsible rows would import styling for this table. We could provide a proxy layer for now and include all current extensions in `@patternfly/react-table` with warning that in future breaking release consumers will have to import these extensions from different package. We'll start developing new extensions only in this package and slowly migrate to more granular approach.

### Why having 3 packages for tables?

So application developers can decide what is best for them and choose one of 3 options: out of the box data driven tables; plain HTML tables; extensions for data driven tables.

### Usage of `@patternfly/react-table-lightweight`

```JSX
import React from 'react';
import {Table, TableHead, TableBody, TableRow, TableCell} from '@patternfly/react-table-lightweight';

export default ({ rows, columns }) => (
  <Table>
    <TableHead>
      <TableRow>
        {columns.map(((column, key)) => (
          <TableCell key={key}>
            {column}
          </TableCell>
        ))}      
      </TableRow>
    </TableHead>
  <TableBody>
    {rows.map((row, key) => (
      <TableRow>
        {row.map((cell, cellKey) => 
          <TableCell key={cellKey}>
            {cell}
          </TableCell>
        )}
      </TableRow>
    ))}
  </TableBody>
 </Table>
);
```

### Usage of `@patternfly/react-table` with `@patternfly/react-table-extensions`

```JSX
import React from 'react';
import {
    Table,
    TableBody,
    TableHeader
} from '@patternfly/react-table';
import { expandable, selectable } from '@patternfly/react-table-extensions';

export default ({ rows, columns }) => (
  <Table rows={rows} cells={columns} extensions={[ expandable, selectable ]}>
    <TableHeader />
    <TableBody />
  </Table>
)

```

This will allow using just specific extensions, making treeshaking more happy and including just the right styles in DOM instead of adding all styles for tables.

**Comment by evwilkin** [comment]
This is related to the work @jschuler is doing in https://github.com/patternfly/patternfly-react/issues/2358, connecting the issues in case there is anything to be shared between them.

**Comment by jschuler** [comment]
@karelhala I've done some work to create a set of composable components in https://github.com/patternfly/patternfly-react/pull/4950, but I've kept it semi-DRY by exporting them from the same react-table package, so that styling is shared across. The heavyweight Table component will use those same composable components under-the-hood.

**Comment by vconzola** [comment]
@mcarrano It appears the selection function for the tree table isn't working correctly. Selection of a parent should automatically select all of its children and selection of a single child should make the parent selection indeterminate. Neither of these things appears to be happening.

**Comment by nicolethoen** [comment]
There are two approaches which have been considered:

## 1. Modify the ComposableTable to allow the tree structure.
Karel and cloud.redhat.com have an implementation of a TreeTable that uses react-table.
They built a decorator, located here: https://github.com/RedHatInsights/frontend-components/blob/master/packages/components/src/TreeTable/decorator.js
And a row wrapper: https://github.com/RedHatInsights/frontend-components/blob/master/packages/components/src/TreeTable/rowWrapper.js
That, along with some css and help functions is used on their overview page. The code is found here: https://github.com/RedHatInsights/api-frontend/blob/master/src/routes/Overview.js
And you can see it in action: https://cloud.redhat.com/docs/api

Since it uses a decorator, I do believe it’d be easily incorporated into the composable table. So in terms of building the nested rows structure, this may be a quicker option.

It would not come with other TreeView features like search or selection cascading into children.

## 2. Modify the TreeTable to have a Table variant. 
It would change the html elements from TreeViewListItem components to use <tr>. It would also expect columns and data for each column. It will be easier to get all of the appropriate tree behaviors out of the box than it will be in the table example.

I started experimenting with whether or not the TreeView component could be modified to use a table html structure under the hood. By replacing the list with a table, I was able to demonstrate that the basic tree view with all its functionality still exists - even if adjustments still need to be made so that the table can be build recursively and have valid html. Code can be found on this branch: https://github.com/nicolethoen/patternfly-react/tree/modify_tree_view_with_table

Ideally, the TreeView would still use a list - but we could create a variant of the TreeView which uses a table html structure, or create a new TreeTable component which shares much of its functionality with the TreeView. My example begins to clumsily demonstrate how we might start considering adding columns to a TreeView - but nesting and searching will still require some work in this case for the TreeTable to have full functionality.

**Comment by karelhala** [comment]
Thank you @nicolethoen for taking a look at this. I think the first option is the right way. We don't want to include search with the treetable as this should be part of toolbar above table, should it be part of the treeTable then it won't be a table but just different variant of treeView. Plus with the second approach we'd have to reimplement all of the decorators such as cellWidth, sortable, actions column, radio/checkbox select etc. You could even have editable rows with TreeTable or mix TreeTable and compound expandable table.

I don't think it would be wise to calculate the selected amount for consumers, we can and probably should send them what user wants to select, but they have to calculate and apply that to the rows themselves. Their structure of rows can be slightly different, they might want to send an API request upon selection with some columns (not only ID but for instance the selected name etc.). We could help them with some helpers and perhaps check if some children is selected and show `-` at the parent, perhaps check if each children is selected and show tick for the parent. But that's all, I wouldn't select all children upon clicking a checkbox at parent. On such action I would just send `event` and newly selected rows and let consumer deal with what is selected.

As you said the decorator for ComposableTable will be easier and faster to use, that's exactly the reason why previously the ReactTabular library was used so we can enahnce and add decarots on the fly without the need of introducing new components and changing the API. Morover if the styling is not yet done it could (and probably should) just enhance the table component, meaning it could heavily depend on table mobile view and such, no need to import one huge CSS just for treeTable.

**Comment by mcarrano** [comment]
@nicolethoen @karelhala I know that having search and nested row selection were requirements from the Migration Toolkit team. Whether that behavior should be integrated or handled by the consumer outside of the component is a valid question. It's true that building on the Table would save us the need to duplicate a lot of table styling and/or behavior, although I don't think we necessarily need all of this for the tree table (sorting or row expansion don't seem to make much sense, for example).

@mattnolting @mturley would also be interested in hearing your thoughts.

**Comment by karelhala** [comment]
@mcarrano we do have a requirement on sort in crc on TreeTable component. You can see in the live environment how it works. I would strongly recommended not to include search within this component as there could be a pagination above the table.

Yeah the expandable table makes no sense with TreeTable, but the rest table enhancers and decorators would have to been re-implemented for the TreeView variant which would take a lot of effort and duplicate a lot of code.

**Comment by mcarrano** [comment]
@karelhala Looking at the demo, looks like the sort only effects the order of the top level rows then and not the content at lower levels in the tree. Is that right? I hadn't thought of it that way, but I can see that as useful since it doesn't break the hierarchy.

**Comment by nicolethoen** [comment]
i'd imagine the search would be optional

**Comment by mcarrano** [comment]
Yes

**Comment by nicolethoen** [comment]
As it stands - it seems that our first attempt should be to take an approach similar to the one taken by the cloud.redhat.com team. We will modify the Table component so that a new decorator and rowwrapper adds the appropriate PatternFly styles to the TreeTable rows. 
Then, features such as search/filter/cascading checkboxes may need to be wired up and included in a demo.

**Comment by jessiehuff** [comment]
I think following the WAI-ARIA structure makes sense for a11y. When testing their implementation it works well for me. 

The keyboard interaction would be similar to the tree view component and follows the example: right/left arrow keys handling expansion and collapsing and shifting focus between cells. Up/down arrow keys allowing the user to navigate by row. Enter performs default action associated with row or cell that has focus (i.e. navigate to link, toggle expansion, etc.) For full keyboard interaction expectations refer to the keyboard support table on this example.


It should be noted that a roving tabindex should be used here:

tabindex="-1" on tr or td
- Makes the element focusable without including it in the tab sequence of the page.
- All row and gridcell elements are focusable, but only one is included in the tab sequence.

tabindex="0" on tr or td
- Includes the element in the tab sequence.
- Only one row or gridcell in the treegrid has tabindex="0" at a time.
- In this implementation, the first row in the treegrid is included in the tab sequence when the page loads.
- When the user moves focus in the treegrid, the element included in the tab sequence changes to the element with focus as described in the section on[ roving tabindex](https://www.w3.org/TR/wai-aria-practices-1.1/#kbd_roving_tabindex).

NOTE: Due to an error in the ARIA 1.1 specification, the aria-posinset and aria-setsize properties are not supported on row elements. This is being corrected in ARIA 1.2. Consequently, when validating the HTML in this example, errors are generated. (So, in other words, if you see an a11y error related to this, it's because these attributes are fairly new, but it's ok. 🙂 )
